core/kernel/interrupt.c

```diff
@@ -196,3 +196,129 @@ void interrupt_remove_free_handler(struct itr_handler *hdl)
+{
+	return dt_driver_register_provider(fdt, node,
+					   (get_of_device_func)dt_get_itr,
+					   data, DT_DRIVER_INTERRUPT);
+}
+
+/*
+ * Provides an itr_desc reference based on "interrupts" property bindings.
+ * May return TEE_ERROR_DEFER_DRIVER_INIT if parent controller is found but
+ * not yet initialized.
+ */
+static TEE_Result get_legacy_interrupt_by_index(const void *fdt, int node,
+						unsigned int index,
+						struct itr_desc **out_desc)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	const uint32_t *prop = NULL;
+	uint32_t phandle = 0;
+	void *desc = NULL;
+	int pnode = 0;
+	int len = 0;
+
+	prop = fdt_getprop(fdt, node, "interrupts", &len);
+	if (!prop)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	/* Find "interrupt-parent" in node or its parents */
+	pnode = node;
+	prop = fdt_getprop(fdt, pnode, "interrupt-parent", &len);
+
+	while (!prop) {
+		pnode = fdt_parent_offset(fdt, pnode);
+		if (pnode < 0)
+			break;
+
+		prop = fdt_getprop(fdt, pnode, "interrupt-parent", &len);
+		if (!prop && len != -FDT_ERR_NOTFOUND)
+			break;
+	}
+	if (!prop) {
+		DMSG("No interrupt parent for node %s",
+		     fdt_get_name(fdt, node, NULL));
+		return TEE_ERROR_GENERIC;
+	}
+
+	/* "interrupt-parent" provides interrupt controller phandle */
+	phandle = fdt32_to_cpu(prop[0]);
+
+	/* Get interrupt chip/number from phandle and "interrupts" property */
+	res = dt_driver_device_from_node_idx_prop_phandle("interrupts", fdt,
+							  node, index,
+							  DT_DRIVER_INTERRUPT,
+							  phandle, &desc);
+	if (!res)
+		*out_desc = desc;
+
+	return res;
+}
+
+/*
+ * Provides an itr_desc based on "interrupts-extended" property bindings.
+ * May return TEE_ERROR_DEFER_DRIVER_INIT if parent controller is found
+ * but not yet initialized.
+ * With this function, provider is expected to have allocated itr_desc
+ * with malloc() or like.
+ */
+static TEE_Result get_extended_interrupt_by_index(const void *fdt, int node,
+						  unsigned int index,
+						  struct itr_desc **out_desc)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	void *desc = NULL;
+
+	res = dt_driver_device_from_node_idx_prop("interrupts-extended",
+						  fdt, node, index,
+						  DT_DRIVER_INTERRUPT, &desc);
+	if (!res)
+		*out_desc = desc;
+
+	return res;
+}
+
+TEE_Result interrupt_dt_get_by_index(const void *fdt, int node,
+				     unsigned int index, struct itr_chip **chip,
+				     size_t *itr_num)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct itr_desc *desc = NULL;
+
+	assert(chip && itr_num);
+
+	/* "interrupts-extended" takes precedence over "interrupts" */
+	if (fdt_getprop(fdt, node, "interrupts-extended", NULL))
+		res = get_extended_interrupt_by_index(fdt, node, index, &desc);
+	else
+		res = get_legacy_interrupt_by_index(fdt, node, index, &desc);
+
+	assert((!res && desc) || (res && !desc));
+
+	if (!res) {
+		*chip = desc->chip;
+		*itr_num = desc->itr_num;
+
+		/* Balance malloc() or like from itr_dt_get_func callback */
+		free(desc);
```

Comment:
This is odd. Why not allocate this on the stack?

Reply:
Makes sense. I think it's feasible.
`get_of_device_func` function type and `dt_driver_device_from_xxx()` currently expect the reference is an output only ref, but I think it's fine if also used as a input argument. I'll propose something.